### PR TITLE
Xenos can chose a xenomorph or a silo to track

### PR DESCRIPTION
**Pull request by BraveMole**
## About The Pull Request

Replace the xeno leader tracker by a selectable tracker. You can chose what to track in the hive status window. Also you can know what is the closest silo from hive status

## Why It's Good For The Game
Much more usefull than the current tracker, no more wandering on the map to find the damn closest silo, or where is that xenos that called for survivor

## Changelog
:cl:
add: Xenos can chose to track another xenomorph or a silo in hive status
add: Distance to silo in hive status
/:cl:

### CODE REVIEW

**Comment by neohenry** [comment]
yay more beno QoL

**Comment by Windryn** [comment]
oh my god **YEEES**

File: code/modules/mob/living/carbon/xenomorph/xenoprocs.dm [hunk 1]
```diff
@@ -39,17 +39,14 @@
...
 	var/leadprefix = (ignore_leads?"":"<b>(-L-)</b>")
 	for(var/i in xenolist)
 		var/mob/living/carbon/xenomorph/X = i
 		if(ignore_leads && X.queen_chosen_lead)
 			continue
-		if(can_overwatch)
-			xenoinfo += "<tr><td>[leadprefix]<a href='byond://?src=\ref[user];watch_xeno_name=[X.nicknumber]'>[X.name]</a> "
-		else
-			xenoinfo += "<tr><td>[leadprefix][X.name] "
+		xenoinfo += "<tr><td>[leadprefix]<a href='byond://?src=\ref[user];track_xeno_name=[X.nicknumber]'>[X.name]</a> "
```
Comment (SplinterGP) [inline code/modules/mob/living/carbon/xenomorph/xenoprocs.dm:49]: doenst this fucks with queen overwatch ? its aparently removing it ?

File: code/modules/mob/living/carbon/xenomorph/castes/queen/queen.dm [hunk 2]
```diff
@@ -93,29 +93,6 @@
 	if(mind)
 		mind.name = name
 
-// ***************************************
-// *********** Icon
-// ***************************************
-/mob/living/carbon/xenomorph/queen/Topic(href, href_list)
```
Comment (SplinterGP) [inline code/modules/mob/living/carbon/xenomorph/castes/queen/queen.dm:99]: same here, doenst this make queen lose overwatch ?

File: code/modules/mob/living/carbon/xenomorph/xenoprocs.dm (hunk 1 above)
Comment (BraveMole) [reply code/modules/mob/living/carbon/xenomorph/xenoprocs.dm:49]: currently, queen can overwatch from two menus
the watch xenos menu
and the hive status menu
so it's just removing one access to overwatch
the action is still there

File: code/modules/mob/living/carbon/xenomorph/xeno_defines.dm [hunk 3]
```diff
@@ -310,4 +310,6 @@
 
 	var/fire_luminosity = 0 //Luminosity of the current fire while burning
 
+	var/tracked
+
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/xeno_defines.dm:314]: New variables must be documented

File: code/modules/mob/living/carbon/xenomorph/castes/queen/queen.dm (hunk 2 above)
Comment (TiviPlus) [reply code/modules/mob/living/carbon/xenomorph/castes/queen/queen.dm:99]: Its inherited from /xenomorph @SplinterGP

File: code/modules/mob/living/carbon/xenomorph/xenomorph.dm [hunk 4]
```diff
@@ -302,25 +302,68 @@
...
+			else
+				var/area/A = get_area(src.loc)
+				var/area/QA = get_area(xeno_tracked.loc)
+				if(A.fake_zlevel == QA.fake_zlevel)
+					LL_dir.icon_state = "trackon"
+					LL_dir.setDir(get_dir(src, xeno_tracked))
+					return
+				else
+					LL_dir.icon_state = "trackondirect"	
+					return
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/xenomorph.dm:328]: excess elses here, you dont need them if you already early return

File: code/modules/mob/living/carbon/xenomorph/xenomorph.dm [hunk 5]
```diff
@@ -302,25 +302,68 @@
...
+
+		else if (istype(tracked, /obj/structure/resin/silo))
+			var/mob/living/carbon/xenomorph/silo_tracked = tracked
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/xenomorph.dm:331]: Not binding since im unsure, but dont we have a isresinsilo

**Review by TiviPlus — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/carbon/xenomorph/xenomorph.dm (hunk 4 above)
Comment (BraveMole) [reply code/modules/mob/living/carbon/xenomorph/xenomorph.dm:328]: Of course, but it is still better from a readability point

File: code/modules/mob/living/carbon/xenomorph/xenomorph.dm (hunk 5 above)
Comment (BraveMole) [reply code/modules/mob/living/carbon/xenomorph/xenomorph.dm:331]: no we don't, just checked

Comment (BraveMole) [reply code/modules/mob/living/carbon/xenomorph/xenomorph.dm:331]: I added it

File: code/__DEFINES/is_helpers.dm [hunk 6]
```diff
@@ -104,6 +104,9 @@
 #define isxenobull(A) (istype(A, /mob/living/carbon/xenomorph/bull))
 #define isxenohivemind(A) (istype(A, /mob/living/carbon/xenomorph/hivemind))
 
+#define isxenohive(A) ((A == XENO_HIVE_NONE) || (A == XENO_HIVE_NORMAL) || (A == XENO_HIVE_CORRUPTED) || (A == XENO_HIVE_ALPHA) || (A == XENO_HIVE_BETA) || (A == XENO_HIVE_ZETA) || (A == XENO_HIVE_ADMEME))
```
Comment (TiviPlus) [inline code/__DEFINES/is_helpers.dm:107]: Considering this isnt a strict istype this isnt the best place to put this

File: code/modules/mob/living/carbon/xenomorph/xeno_defines.dm [hunk 7]
```diff
@@ -313,5 +313,8 @@
 	var/notice_delay = 20 //2 second between notices
 
 	var/fire_luminosity = 0 //Luminosity of the current fire while burning
+	
+	///The xenos/silo currently tracked by the xeno_tracker arrow
+	var/tracked
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/xeno_defines.dm:318]: Does this have protection against hard dels?

File: code/modules/mob/living/carbon/xenomorph/xenomorph.dm [hunk 8]
```diff
@@ -302,25 +302,50 @@
...
+	
+	if (isxeno(tracked))
+		var/mob/living/carbon/xenomorph/xeno_tracked = tracked
+		if(xeno_tracked == src) // No need to track ourselves
+			LL_dir.icon_state = "trackoff"
+			return	
+		if(xeno_tracked.z != src.z || get_dist(src,xeno_tracked) < 1)
+			LL_dir.icon_state = "trackondirect"
+			return
+		var/area/A = get_area(src.loc)
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/xenomorph.dm:324]: src.var 
>:(

File: code/modules/mob/living/carbon/xenomorph/xenomorph.dm [hunk 9]
```diff
@@ -302,25 +302,50 @@
...
-		LL_dir.icon_state = "trackondirect"
-	else
+	if (isresinsilo(tracked))
+		var/mob/living/carbon/xenomorph/silo_tracked = tracked
+
+		if(silo_tracked.z != src.z || get_dist(src,silo_tracked) < 1)
+			LL_dir.icon_state = "trackondirect"
+			return
+	
 		var/area/A = get_area(src.loc)
```
Comment (TiviPlus) [inline code/modules/mob/living/carbon/xenomorph/xenomorph.dm:341]: ditto

**Review by TiviPlus — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/carbon/xenomorph/xenomorph.dm (hunk 8 above)
Comment (BraveMole) [reply code/modules/mob/living/carbon/xenomorph/xenomorph.dm:324]: One day i'll stop copy paste thing, i promiss

**Review by TiviPlus — APPROVED** [review]

**Pull request merged**